### PR TITLE
Feature - Ignore window decoration for certain views matched by criteria

### LINKED ISSUE #563: Support for Floating windows in Window Rules plugin
Migrating to Wayfire from Sway, I was stuck on getting floating windows to work. I use them for minimal scripts, which uses fzf in a floating window. I thought there would be support in Wayfire but I haven't been able to find anything as such neither in wiki nor in window rules code.

[Here's the Archwiki](https://wiki.archlinux.org/index.php/Sway#Floating_windows) doc on Floating windows in Sway.
Can you please help me out if it already exists or if I can help contribute a fix for it?

Thanks for the awesome WM though! Loved the unique experience :smile:

### PR DESCRIPTION
**Pull request by shivanshs9**
Fixes #563 

- [x] Tested

### CODE REVIEW

**Comment by ammen99** [comment]
Hi, thanks for the patch, but the correct place for checking whether to have a decoration or not is here: https://github.com/WayfireWM/wayfire/blob/master/plugins/decor/decoration.cpp#L41

`deco-theme`'s responsibility is how to draw the decorations, not whether they are drawn at all.

**Comment by shivanshs9** [comment]
Is it possible to disable client-side decorations too?

**Comment by shivanshs9** [comment]
BUMP.

> I also wanted to know how can I disable client-side decorations too? Since I haven't been able to achieve the same for Alacritty since it uses its own decorations.

@ammen99 any ideas for it?

File: .gitignore [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+/subprojects/*
+!/subprojects/*.wrap
+
+### Ninja ###
+.ninja_deps
+.ninja_log
+
+### VisualStudioCode ###
+.vscode/*
+!.vscode/settings.json
```
Comment (ammen99) [inline .gitignore:65]: We do not use vscode, so I do not think we need these exceptions in the .gitignore ..

**Comment by ammen99** [comment]
LGTM aside from the comment.



> @ammen99 any ideas for it?

Try setting https://github.com/WayfireWM/wayfire/wiki/Configuration#corepreferred_decoration_mode to `server`, and then tell `alacritty` to use SSD. Finally, add an exception for alacritty in the decoration plugin (via the option you have implemented in this PR).

**Comment by ammen99** [comment]
Oh, I just remembered this comment here: https://github.com/WayfireWM/wayfire/issues/551#issuecomment-640328958, maybe it will be helpful too.

File: .gitignore (hunk 1 above)
Comment (shivanshs9) [reply .gitignore]: Fixed.

**Comment by shivanshs9** [comment]
> Try setting [WayfireWM/wayfire/wiki/Configuration#corepreferred_decoration_mode](https://github.com/WayfireWM/wayfire/wiki/Configuration#corepreferred_decoration_mode) to `server`, and then tell `alacritty` to use SSD. Finally, add an exception for alacritty in the decoration plugin (via the option you have implemented in this PR).

Ah, I've already been using it this way but I felt it's like a hack since it sets server side decoration globally, when the aim of this plugin was simply to disable window decoration of windows matched by criteria. But I understand if there's no other way to do it without setting it globally.

> Oh, I just remembered this comment here: #551 (comment), maybe it will be helpful too.

Yep I've been using the same config.

**Comment by ammen99** [comment]
> But I understand if there's no other way to do it without setting it globally.

Theory says it is possible to extend the option in core to match preferred mode per view. However, this is complicated by the fact that xdg-decoration objects are per-surface and the initial negotiation happens before the toplevel is mapped, which means that at least initially we cannot use matcher to separate views from each other. One would have to wait until map event and change things then ...

**Comment by ammen99** [comment]
Can you squash your 1th and 4th commits and the 2nd and 3rd?

**Comment by shivanshs9** [comment]
> Can you squash your 1st and 4th commits and the 2nd and 3rd?

Yep done! :+1: 

> Theory says it is possible to extend the option in core to match preferred mode per view. However, this is complicated by the fact that xdg-decoration objects are per-surface and the initial negotiation happens before the toplevel is mapped, which means that at least initially we cannot use matcher to separate views from each other. One would have to wait until map event and change things then ...

Got it, I'll go with the hack for now then! :sweat_smile:

**Comment by ammen99** [comment]
Thanks!

**Pull request merged**